### PR TITLE
Pgi get mesh proxy

### PR DESCRIPTION
**Pull request by christophermaynard**
Modified code and tests. The tests all pass!

### CODE REVIEW

File: config/psyclone.cfg [hunk 1]
```diff
@@ -59,8 +59,7 @@ access_mapping = gh_read:read, gh_write: write, gh_rw: readwrite,
 # remove the need for a halo exchange call.
 COMPUTE_ANNEXED_DOFS = false
 
-access_mapping = gh_read:read, gh_write: write, gh_readwrite: readwrite,
-                 gh_inc: inc, gh_sum: sum
+access_mapping = gh_read:read, gh_write: write, gh_readwrite: readwrite, gh_inc: inc, gh_sum: sum
```
Comment (arporter) [inline config/psyclone.cfg]: Is this change needed? If not, please undo!

File: src/psyclone/dynamo0p3.py [hunk 2]
```diff
@@ -3105,13 +3106,16 @@ def initialise(self, parent):
                 parent.add(
                     AssignGen(parent, pointer=True,
                               lhs=fine_mesh,
-                              rhs=dig.fine.name_indexed + "%get_mesh()"))
+                              rhs="%".join([dig.fine.proxy_name_indexed,
+                                                dig.fine.ref_name(),"get_mesh()"]) ))
```
Comment (arporter) [inline src/psyclone/dynamo0p3.py]: Lots of pycodestyle errors here!

File: src/psyclone/dynamo0p3.py [hunk 3]
```diff
@@ -3105,13 +3106,16 @@ def initialise(self, parent):
...
+
             if coarse_mesh not in initialised:
                 initialised.append(coarse_mesh)
                 parent.add(
                     AssignGen(parent, pointer=True,
                               lhs=coarse_mesh,
-                              rhs=dig.coarse.name_indexed + "%get_mesh()"))
+                              rhs="%".join([dig.coarse.proxy_name_indexed,
```
Comment (arporter) [inline src/psyclone/dynamo0p3.py]: pycodestyle errors here too.

**Comment by christophermaynard** [comment]
Ready for first review from @arporter.

File: config/psyclone.cfg (hunk 1 above)
Comment (christophermaynard) [reply config/psyclone.cfg]: Oops. Reverted.

File: src/psyclone/dynamo0p3.py (hunk 2 above)
Comment (christophermaynard) [reply src/psyclone/dynamo0p3.py]: Fixed.

File: src/psyclone/dynamo0p3.py (hunk 3 above)
Comment (christophermaynard) [reply src/psyclone/dynamo0p3.py]: Fixed.

**Comment by christophermaynard** [comment]
Hi @arporter, I have made the specific changes requested. However, I am not sure what you want me to do and I don't know how to do it for the _significant issue_. I am trying to find and example or documentation. Thanks.

**Comment by arporter** [comment]
Hi @christophermaynard, if you search for "compile" in e.g. `dynamo0p3_test.py` you'll find lines/calls that test that the Fortran generated in the test actually compiles, e.g.:

    assert Dynamo0p3Build(tmpdir).code_compiles(psy)

 Please could you add similar lines to all of the tests you've modified.

**Comment by arporter** [comment]
Forgot to say: you also need to check that the compilation tests pass by giving the `--compile` flag to py.test

**Comment by arporter** [comment]
Just remembered: you also kindly volunteered to fix the error in the Developers' Guide :-)

**Comment by christophermaynard** [comment]
Hi @arporter. I have fixed all the compile tests. Not sure what should happen now in the workflow. Can I just give it back to you?

**Comment by christophermaynard** [comment]
... and the documentation.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 4]
```diff
@@ -531,7 +531,7 @@ def test_field(tmpdir):
     assert output in str(generated_code)
 
 
-def test_field_deref(dist_mem):
+def test_field_deref(tmpdir,dist_mem):
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py]: pycodestyle error here.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 5]
```diff
@@ -3141,9 +3159,6 @@ def test_halo_dirty_4():
         "      !\n"
         "      ! Set halos dirty/clean for fields modified in the above loop\n"
         "      !\n"
-        "      CALL chi_proxy(1)%set_dirty()\n"
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py]: I don't think halo exchanges should be affected by this PR. Please could you investigate what's happened to them.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 6]
```diff
@@ -3469,7 +3484,7 @@ def test_fs_anyspace_and_readwrite_error():
                 in str(excinfo.value))
 
 
-def test_halo_exchange_view(capsys):
+def test_halo_exchange_view(tmpdir,capsys):
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py]: pycodestyle error. In fact, it looks as though `tmpdir` isn't used in this test?

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 [hunk 7]
```diff
@@ -14,12 +14,10 @@ program single_invoke
...
   type(field_type) :: f1, f2, m1, m2
-  real(r_def)      :: a
-  integer(i_def)   :: iflag, istep
+  integer(i_def)   :: iflag, istep, iscalar
 
   call invoke(                                  &
        testkern_type(iflag,f1,f2,m1,m2,istep),  &
-       testkern_type(1_i_def,f1,f2,m1,m2,istep) &
-          )
+       testkern_type(iflag,f1,f2,m1,m2,iscalar)   )
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: This test was explicitly checking the case where an argument is passed by value. Please could you revert this part of the change (and the associated pytest test of course).

File: src/psyclone/tests/test_files/dynamo0p3/testkern_chi_2.F90 [hunk 8]
```diff
@@ -6,18 +6,30 @@
 !-------------------------------------------------------------------------------
 ! Author R. Ford STFC Daresbury Lab
 
-module testkern_chi_mod
-  type, extends(kernel_type) :: testkern_chi_type
+module testkern_chi_2
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_chi_2.F90]: You can add yourself to the author list now :-)

File: src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90 [hunk 9]
```diff
@@ -0,0 +1,45 @@
+!-------------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90]: Please update this copyright notice to the one used in other files.

Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90]: Oh, and if you rename a file, it's good if you do it with `git mv oldfile newfile` as then git knows about it.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_two_int_scalars.f90 [hunk 10]
```diff
@@ -6,7 +6,12 @@
 !-------------------------------------------------------------------------------
 ! Author R. Ford STFC Daresbury Lab
 
-module testkern
+module testkern_two_int_scalars
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_two_int_scalars.f90]: Please add yourself to the author list and update the copyright notice/date if necessary.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_two_real_scalars.f90 [hunk 11]
```diff
@@ -6,7 +6,11 @@
 !-------------------------------------------------------------------------------
 ! Author R. Ford STFC Daresbury Lab
 
-module testkern
+module testkern_two_real_scalars
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_two_real_scalars.f90]: Author list and copyright notice please.

File: config/psyclone.cfg [hunk 12]
```diff
@@ -60,7 +60,7 @@ access_mapping = gh_read:read, gh_write: write, gh_rw: readwrite,
 COMPUTE_ANNEXED_DOFS = false
 
 access_mapping = gh_read:read, gh_write: write, gh_readwrite: readwrite,
-                 gh_inc: inc, gh_sum: sum
+	         gh_inc: inc, gh_sum: sum
```
Comment (arporter) [inline config/psyclone.cfg]: Please undo this whitespace change.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 5 above)
Comment (christophermaynard) [reply src/psyclone/tests/dynamo0p3_test.py]: Hi @arporter. I removed them because chi is the coordinate field and is static, so never needs a halo exchange call on it. However, thinking about it, that is not what you are testing here, merely can Psyclone generate a halo exchange on a 3-field. I am reasonable confident that we only use the 3-field like this for static fields. I'll reinstate the tests.

Comment (christophermaynard) [reply src/psyclone/tests/dynamo0p3_test.py]: Done.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 4 above)
Comment (christophermaynard) [reply src/psyclone/tests/dynamo0p3_test.py]: Done.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 6 above)
Comment (christophermaynard) [reply src/psyclone/tests/dynamo0p3_test.py]: Removed.

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 (hunk 7 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: This one is a bit harder @arporter. PSyclone doesn't generate the PSy layer with `i_def`. Not in the module use statement, nor in any of the declarations. The generated algorithm code doesn't have the literal constant passed to the PSy layer. It only appears as an argument to the kernel call in the PSy layer. So, I can't have a test for when an argument is passed in as a literal constant, because it PSyclone doesn't generate the correct code.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_chi_2.F90 (hunk 8 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_chi_2.F90]: Done.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90 (hunk 9 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90]: Which copyright notice? They all have different ones.
Regarding the second comment, I assume that it is just that, a comment? Do _this_ next time?

File: src/psyclone/tests/test_files/dynamo0p3/testkern_two_int_scalars.f90 (hunk 10 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_two_int_scalars.f90]: Added myself as an author, but which copyright notice?

File: config/psyclone.cfg (hunk 12 above)
Comment (christophermaynard) [reply config/psyclone.cfg]: Done.

Comment (arporter) [reply config/psyclone.cfg]: It's still showing as a change in github!

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 [hunk 13]
```diff
@@ -1,25 +1,52 @@
...
   type(field_type) :: f1, f2, m1, m2
-  real(r_def)      :: a
-  integer(i_def)   :: iflag, istep
+  integer(i_def)   :: iflag, istep, iscalar
 
   call invoke(                                  &
        testkern_type(iflag,f1,f2,m1,m2,istep),  &
-       testkern_type(1_i_def,f1,f2,m1,m2,istep) &
-          )
+       testkern_type(istep,f1,f2,m1,m2,iscalar)   )
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: Please put back the integer literal in this test (albeit without the `_i_def`)

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 [hunk 14]
```diff
@@ -1,25 +1,52 @@
...
   ! Description: kernel that has two integer, scalar arguments
   ! specified in an invoke call. One called by value.
   use testkern_two_int_scalars, only: testkern_type
   use inf,      only: field_type
+  use constants_mod, only: r_def, i_def
   implicit none
   type(field_type) :: f1, f2, m1, m2
-  real(r_def)      :: a
-  integer(i_def)   :: iflag, istep
+  integer(i_def)   :: iflag, istep, iscalar
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: Please rm `iscalar` from this list if it's not used.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90 (hunk 9 above)
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_one_int_scalar_mod.f90]: Yes, that was for future reference :-)

File: src/psyclone/tests/test_files/dynamo0p3/1.6.3_single_invoke_multiple_derived_types.f90 [hunk 15]
```diff
@@ -11,17 +11,17 @@ program single_invoke
   ! Description: four kernels specified in an invoke call where the
   ! one integer argument is obtained from different components of two
   ! different objects
-  use testkern_one_int_scalar, only: testkern_type
+  use testkern_one_int_scalar_mod, only: testkern_one_int_scalar_type
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6.3_single_invoke_multiple_derived_types.f90]: Please update the copyright notice in this file.

File: src/psyclone/tests/test_files/dynamo0p3/1.6.2_single_invoke_1_int_from_derived_type.f90 [hunk 16]
```diff
@@ -14,7 +14,7 @@ program single_invoke
   ! In the third the type-bound routine takes an argument and in the
   ! fourth this argument is itself obtained by dereferencing another
   ! derived type.
-  use testkern_one_int_scalar, only: testkern_type
+  use testkern_one_int_scalar_mod, only: testkern_one_int_scalar_type
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6.2_single_invoke_1_int_from_derived_type.f90]: Please update the copyright notice in this file.

File: src/psyclone/tests/test_files/dynamo0p3/1.6.1_single_invoke_1_int_scalar.f90 [hunk 17]
```diff
@@ -9,14 +9,14 @@
 program single_invoke
 
   ! Description: single function specified in an invoke call
-  use testkern_one_int_scalar, only: testkern_type
+  use testkern_one_int_scalar_mod, only: testkern_one_int_scalar_type
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/1.6.1_single_invoke_1_int_scalar.f90]: Please update the copyright notice in this file.

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 (hunk 13 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: Hi @arporter 
We discussed this and the test doesn't work. I have made an issue (#447) to fix it.

Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: I humbly submit that the test does work - I've just tried it. Please revert this change (i.e. replace `istep` with `1` rather than `1_i_def`).

Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: Yes it does actually work (to my surprise) but I claim it tests the wrong thing. I'll make the change.

File: src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90 (hunk 14 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6_single_invoke_2_int_scalars.f90]: Done.

File: src/psyclone/tests/test_files/dynamo0p3/1.6.3_single_invoke_multiple_derived_types.f90 (hunk 15 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6.3_single_invoke_multiple_derived_types.f90]: Done

File: src/psyclone/tests/test_files/dynamo0p3/1.6.1_single_invoke_1_int_scalar.f90 (hunk 17 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6.1_single_invoke_1_int_scalar.f90]: Done.

File: src/psyclone/tests/test_files/dynamo0p3/1.6.2_single_invoke_1_int_from_derived_type.f90 (hunk 16 above)
Comment (christophermaynard) [reply src/psyclone/tests/test_files/dynamo0p3/1.6.2_single_invoke_1_int_from_derived_type.f90]: Done.

**Comment by christophermaynard** [comment]
Hi @arporter. Hopefully we have now caught everything except possibly the white space change. I admit that it is rather annoying to have white space change committed to trunk but I can't seem to undo it.
